src/offload_gemm.F

```diff
@@ -0,0 +1,395 @@
+
+   INTEGER, PARAMETER, PUBLIC :: &
+      OFFLOAD_GEMM_PU_HOST = 0, &
+      OFFLOAD_GEMM_PU_GPU = 1
+CONTAINS
+
+! **************************************************************************************************
+!> \brief ...
+!> \param opA ...
+!> \param opB ...
+!> \param m ...
+!> \param n ...
+!> \param k ...
+!> \param alpha ...
+!> \param A ...
+!> \param lda ...
+!> \param B ...
+!> \param ldb ...
+!> \param beta ...
+!> \param C ...
+!> \param ldc ...
+!> \param ctx ...
+! **************************************************************************************************
+   SUBROUTINE offload_dgemm(opA, opB, m, n, k, &
+                            alpha, A, lda, B, ldb, &
+                            beta, C, ldc, ctx)
+      CHARACTER, INTENT(in) :: opA
+      CHARACTER, INTENT(in) :: opB
+      INTEGER, INTENT(in) :: m
+      INTEGER, INTENT(in) :: n
+      INTEGER, INTENT(in) :: k
+      REAL(8), INTENT(in) :: alpha
+      REAL(8), DIMENSION(*), INTENT(in), TARGET :: A
+      INTEGER, INTENT(in) :: lda
+      REAL(8), DIMENSION(*), INTENT(in), TARGET :: B
+      INTEGER, INTENT(in) :: ldb
+      REAL(8), INTENT(in) :: beta
+      REAL(8), DIMENSION(*), INTENT(inout), TARGET ::C
+      INTEGER, INTENT(in) :: ldc
+      TYPE(C_ptr), OPTIONAL, INTENT(inout) :: ctx
+      ! no point of using SPLA offloading on CPU ONLY nodes
+#if defined(__SPLA)
+      INTEGER :: spla_op_A, spla_op_B, spla_error
+
+      IF (PRESENT(ctx)) THEN
+
+         IF (opA == 'N') spla_op_A = SPLA_OP_NONE
+         IF (opA == 'T') spla_op_A = SPLA_OP_TRANSPOSE
+
+         IF (opB == 'N') spla_op_B = SPLA_OP_NONE
+         IF (opB == 'T') spla_op_B = SPLA_OP_TRANSPOSE
+
+#if __GNUC__ >= 9
+         CPASSERT(IS_CONTIGUOUS(A))
+         CPASSERT(IS_CONTIGUOUS(B))
+         CPASSERT(IS_CONTIGUOUS(C))
+#endif
+
+         spla_error = spla_dgemm(spla_op_A, spla_op_B, &
+                                 m, n, k, alpha, &
+                                 c_loc(A), lda, &
+                                 c_loc(B), ldb, &
+                                 beta, c_loc(C), ldc, ctx)
+         CPASSERT(spla_error == SPLA_SUCCESS)
+      ELSE
+#endif
+         CALL dgemm(opA, opB, m, n, k, alpha, &
+                    A, lda, &
+                    B, ldb, beta, C, ldc)
+#if defined(__SPLA)
+      END IF
+#else
+      MARK_USED(ctx)
+#endif
+   END SUBROUTINE offload_dgemm
+
+! **************************************************************************************************
+!> \brief offload sgemm to GPU when beneficial. All arguemnts identical to blas sgemm
+!> \param opA ...
+!> \param opB ...
+!> \param m ...
+!> \param n ...
+!> \param k ...
+!> \param alpha ...
+!> \param A ...
+!> \param lda ...
+!> \param B ...
+!> \param ldb ...
+!> \param beta ...
+!> \param C ...
+!> \param ldc ...
+!> \param ctx ...
+! **************************************************************************************************
+   SUBROUTINE offload_sgemm(opA, opB, m, n, k, &
```

Comment:
These other `gemm` routines are not being used and probably never will. So, I'd rather remove them.